contracts/OffsetHelper.sol

```diff
@@ -146,6 +146,48 @@ contract OffsetHelper is OffsetHelperStorage {
         autoRetire(tco2s, amounts);
     }
 
+    /**
+     * @notice Retire carbon credits using the lowest quality (oldest) TCO2
+     * tokens available from the specified Toucan token pool by sending ERC20
+     * tokens (USDC, WETH, WMATIC). All provided token is consumed for
+     * offsetting.
+     *
+     * This function:
+     * 1. Swaps the ERC20 token sent to the contract for the specified pool token.
+     * 2. Redeems the pool token for the poorest quality TCO2 tokens available.
+     * 3. Retires the TCO2 tokens.
+     *
+     * Note: The client must approve the ERC20 token that is sent to the contract.
+     *
+     * @dev When automatically redeeming pool tokens for the lowest quality
+     * TCO2s there are no fees and you receive exactly 1 TCO2 token for 1 pool
+     * token.
+     *
+     * @param _fromToken The address of the ERC20 token that the user sends
+     * (must be one of USDC, WETH, WMATIC)
+     * @param _amountToSwap The amount of ERC20 token to swap into Toucan pool
+     * token. Full amount will be used for offsetting.
+     * @param _poolToken The address of the Toucan pool token that the
+     * user wants to use, for example, NCT or BCT
+     *
+     * @return tco2s An array of the TCO2 addresses that were redeemed
+     * @return amounts An array of the amounts of each TCO2 that were redeemed
+     */
+    function autoOffsetUsingToken(
```

Comment:
I'm not saying a hard no, but I'm not a big fan of overloads tbh. The initial `OffsetHelper` had the (I think) and we eventually went to using different method names.

It's also important to think about the fact that using overloads (or renaming existent methods too) may break code out there 🤔

Reply:
Glad you propose this, I'm also not a fan of overloads as they can just cause confusion. The reason I decided for overloads is that I saw you using overloads for the public function `swap`, so I figured it would match the used style.

So what do you think about using the names `autoOffsetUsingFixedToken` and `autoOffsetUsingFixedETH`?

Do you want to keep the `swap` overloads, or also rename all `swap` functions to e.g.
- `swapToken`,
- `swapFixedToken`,
- `swapETH` and
- `swapFixedETH`
? Even though it might potentially break code out there, it would be more consistent with the requirement of not using overloads, especially for public functions. Is the API really fixed already?

Reply:
> The reason I decided for overloads is that I saw you using overloads for the public function swap, so I figured it would match the used style.

I see your point. I think I might do a follow-up PR to move away from overloads for that too.

> using the names `autoOffsetUsingFixedToken` and `autoOffsetUsingFixedETH`?

I like it but can we do `autoOffsetUsingFixedInToken` and `autoOffsetUsingFixedInETH`? My reasoning for using `FixedIn` is related to the one in this [comment](https://github.com/ToucanProtocol/OffsetHelper/pull/4#discussion_r1022954521) I made yesterday.

> Even though it might potentially break code out there

I thought about it, and while it's definitely not a good thing, it's also not the worst thing. This is a yet-to-be-audited contract, so I guess ppl should expect changes. Also, since it's not an upgradeable contract, we will not break anybody's code, unless they specifically want to use the latest deployment.

So, yeah, I'm for renaming `swap` to `swapFixedOutToken`, `swapedFixedInToken`, `swapFixedOutETH`, and `swapFixedInETH`.

Or do you think that maybe my `FixedIn` vs `FixedOut` idea may turn everything too verbose? And we should instead stick to using the `fixed` keyword when doing "fixed in" and no keyword when doing "fixed out"? 🤔

Reply:
Or instead of `Fixed` better use `Exact` everywhere to be in line with Uniswap nomenclature?

Reply:
I like your last proposal to use `FixedIn` and `FixedOut`. Even though it is a bit more verbose, it is much clearer! Maybe now that we change the names, we can also try to be in line with Uniswap and use `ExactIn` and `ExactOut`? And while we're at it, we can remove the `Using` from the `autoOffset` function, so final function names would be
- `swapExactOutToken`
- `swapExactInToken`
- `swapExactOutETH`
- `swapExactInETH`
and
- `autoOffsetExactOutToken`
- `autoOffsetExactInToken`
- `autoOffsetExactOutETH`
- `autoOffsetExactInETH`

What do you think?

Reply:
I love everything you just said 💪🏻 LFG

Reply:
Done 🚀 also took the liberty to remove the `Using` from `autoOffsetUsingPoolToken` to stay consistent. And also removed the overloads for function `calculateExpectedPoolTokenAmount` by renaming them to `calculateExpectedPoolTokenForETH` and `calculateExpectedPoolTokenForToken`.